avogadro/core/cube.cpp

```diff
@@ -193,9 +201,159 @@ float Cube::value(int i, int j, int k) const
+    std::array<std::array<float, 2>, 3> x;
+    std::array<float, 3> run;
+
+    // X-direction
+    if (i == 0) {
+        x[0][0] = m_data[dataIdx + ny * nz]; // (i+1, j, k)
+        x[0][1] = m_data[dataIdx];           // (i, j, k)
+        run[0] = m_spacing.x();
+    } else if (i == (nx - 1)) {
+        x[0][0] = m_data[dataIdx];             // (i, j, k)
+        x[0][1] = m_data[dataIdx - ny * nz];   // (i-1, j, k)
+        run[0] = m_spacing.x();
+    } else {
+        x[0][0] = m_data[dataIdx + ny * nz];    // (i+1, j, k)
+        x[0][1] = m_data[dataIdx - ny * nz];    // (i-1, j, k)
+        run[0] = 2 * m_spacing.x();
+    }
+
+    // Y-direction
+    if (j == 0) {
+        x[1][0] = m_data[dataIdx + nz];   // (i, j+1, k)
+        x[1][1] = m_data[dataIdx];        // (i, j, k)
+        run[1] = m_spacing.y();
+    } else if (j == (ny - 1)) {
+        x[1][0] = m_data[dataIdx];        // (i, j, k)
+        x[1][1] = m_data[dataIdx - nz];   // (i, j-1, k)
+        run[1] = m_spacing.y();
+    } else {
+        x[1][0] = m_data[dataIdx + nz];    // (i, j+1, k)
+        x[1][1] = m_data[dataIdx - nz];    // (i, j-1, k)
+        run[1] = 2 * m_spacing.y();
+    }
+
+    // Z-direction
+    if (k == 0) {
+        x[2][0] = m_data[dataIdx + 1];     // (i, j, k+1)
+        x[2][1] = m_data[dataIdx];         // (i, j, k)
+        run[2] = m_spacing.z();
+    } else if (k == (nz - 1)) {
+        x[2][0] = m_data[dataIdx];         // (i, j, k)
+        x[2][1] = m_data[dataIdx - 1];     // (i, j, k-1)
+        run[2] = m_spacing.z();
+    } else {
+        x[2][0] = m_data[dataIdx + 1];     // (i, j, k+1)
+        x[2][1] = m_data[dataIdx - 1];     // (i, j, k-1)
+        run[2] = 2 * m_spacing.z();
+    }
+
+    std::array<float, 3> ret;
+
+    ret[0] = (x[0][1] - x[0][0]) / run[0];
+    ret[1] = (x[1][1] - x[1][0]) / run[1];
+    ret[2] = (x[2][1] - x[2][0]) / run[2];
+
+    return ret;
+}
+
+std::array<std::array<float, 3>, 8>
+Cube::getGradCube(int i, int j, int k) const
 {
+    std::array<std::array<float, 3>, 8> grad;
+
+    grad[0] = computeGradient(i, j, k);
+    grad[1] = computeGradient(i + 1, j, k);
+    grad[2] = computeGradient(i + 1, j + 1, k);
+    grad[3] = computeGradient(i, j + 1, k);
+    grad[4] = computeGradient(i, j, k + 1);
+    grad[5] = computeGradient(i + 1, j, k + 1);
+    grad[6] = computeGradient(i + 1, j + 1, k + 1);
+    grad[7] = computeGradient(i, j + 1, k + 1);
+
+    return grad;
+}
+
+std::array<float, 8> Cube::getValsCube(int i, int j, int k) const
+{
+    std::array<float, 8> vals;
+    
+    vals[0] = getData(i, j, k);
+    vals[1] = getData(i + 1, j, k);
+    vals[2] = getData(i + 1, j + 1, k);
+    vals[3] = getData(i, j + 1, k);
+    vals[4] = getData(i, j, k + 1);
+    vals[5] = getData(i + 1, j, k + 1);
+    vals[6] = getData(i + 1, j + 1, k + 1);
+    vals[7] = getData(i, j + 1, k + 1);
+
+    return vals;
+}
+
+
+inline float Cube::getData(int i, int j, int k) const
```

Comment:
Is that a warning or an error? @ghutchis? Because here's the implementation of getData(int, int ,int). 

My code runs and compile correctly. Can you confirm it once?

Reply:
Yes, it's a build error. It's failing on my machine and all build actions.

If I remove the `inline` and assume the compiler will handle it, everything compiles fine.